galsim/base.py

```diff
@@ -338,6 +338,258 @@ def getGSParams(self):
+        @param flux_frac    The fraction of light to be enclosed by the returned radius.
+                            [default: 0.5]
+
+        @returns an estimate of the half-light radius
+        """
+        if hasattr(self, 'half_light_radius'): 
+            return self.half_light_radius
+
+        if scale is None:
+            scale = self.nyquistScale() * 0.5
+
+        if centroid is None:
+            centroid = self.centroid()
+
+        # Draw the image.  Note: need a method that integrates over pixels to get flux right.
+        # The offset is to make all the rsq values different to help the precision a bit.
+        offset = galsim.PositionD(0.2, 0.33)
+        im = self.drawImage(nx=size, ny=size, scale=scale, offset=offset)
+
+        # Use radii at centers of pixels as approximation to the radial integral
+        x,y = np.meshgrid(range(im.array.shape[0]), range(im.array.shape[1]))
+        x = x - (im.trueCenter().x - im.bounds.xmin + offset.x + centroid.x/scale)
+        y = y - (im.trueCenter().y - im.bounds.ymin + offset.y + centroid.y/scale)
+        rsq = x*x + y*y
+
+        # Sort by radius
+        indx = np.argsort(rsq.flatten())
+        rsqf = rsq.flatten()[indx]
+        data = im.array.flatten()[indx]
+        cumflux = np.cumsum(data)
+
+        # Find the first value with cumflux > 0.5 * flux
+        k = np.argmax(cumflux > flux_frac * self.flux)
+        flux_k = cumflux[k] / self.flux  # normalize to unit total flux
+
+        # Interpolate (linearly) between this and the previous value.
+        if k == 0:
+            hlrsq = rsqf[0] * (flux_frac / flux_k)
+        else:
+            fkm1 = cumflux[k-1] / self.flux
+            # For brevity in the next formula:
+            fk = flux_k
+            f = flux_frac
+            hlrsq = (rsqf[k-1] * (fk-f) + rsqf[k] * (f-fkm1)) / (fk-fkm1)
+
+        # This has all been done in pixels.  So normalize according to the pixel scale.
+        hlr = np.sqrt(hlrsq) * im.scale
+
+        return hlr
+
+
+    def calculateMomentRadius(self, size=None, scale=None, centroid=None, rtype='det'):
+        """Returns an estimate of the radius based on second moments.
+
+        The second moments are defines as:
+
+        Q_ij = int( I(x,y) i j dx dy ) / int( I(x,y) dx dy )
+        where i,j may be either x or y.
+
+        If I(x,y) is a Gaussian, then T = Tr(Q) = Qxx + Qyy = 2 sigma^2.  Thus, one reasonable
+        choice for a "radius" for an arbitrary profile is sqrt(T/2).  
+
+        In addition, det(Q) = sigma^4.  So another choice for an arbitrary profile is det(Q)^1/4.
+
+        This routine can return either of these measures according to the value of the `rtype`
+        parameter.  `rtype='trace'` will cause it to return sqrt(T/2).  `rtype='det'` will cause
+        it to return det(Q)^1/4.  And `rtype='both'` will return a tuple with both values.
+
+        Note that for the special case of a Gaussian profile, no calculation is necessary, and
```

Comment:
Well, mathematically the approximation is that we take int_pixel (I(x,y) i j) dx dy ~= (int_pixel I(x,y) dx dy) * i_pixel_center * j_pixel_center.  It's not that we are using a different profile.  It seems to me that this is already implicit in the description of the method.  Especially that it gets more accurate as the pixel scale decreases.  But I'll add this caveat explicitly.